src/Oscoin/Storage/ContentStore.hs

```diff
@@ -0,0 +1,54 @@
+module Oscoin.Storage.ContentStore
+    ( ContentStore
+    , storeContent
+    , lookupContent
+
+    , newIoContentStore
+    , mkStateContentStore
+    , hoistContentStore
+    ) where
+
+import           Oscoin.Prelude
+
+import qualified Oscoin.Crypto.Hash as Crypto
+
+import qualified Data.Map as Map
+import           Lens.Micro
+import           Lens.Micro.Mtl
+
+data ContentStore crypto a m = ContentStore
+    { storeContent  :: a -> m ()
+    , lookupContent :: Crypto.Hashed crypto a -> m (Maybe a)
+    }
+
+hoistContentStore :: (forall x. m x -> n x) -> ContentStore crypto a m -> ContentStore crypto a n
+hoistContentStore lift' c = ContentStore
+    { storeContent = lift' . storeContent c
+    , lookupContent = lift' . lookupContent c
+    }
+
+newIoContentStore
+    :: forall crypto a m.
+       ( MonadIO m
+       , Crypto.Hashable crypto a)
+    => IO (ContentStore crypto a m)
+newIoContentStore = do
+    mapVar <- newMVar mempty
+    let lookupContent hash = liftIO $ readMVar mapVar <&> Map.lookup hash
+    let storeContent a = liftIO $ modifyMVar_ mapVar $ \m -> pure $ Map.insert (Crypto.hash @crypto a) a m
```

Comment:
Should we worry about inserting/replacing the content? Do we need to keep a "list of previous contents" we can use for looking things up "in the past"?

Reply:
I don’t think we need this. The keys is the content’s hash so we cannot replace content strictly speaking. If two values have the same hash then they are the same.

Reply:
> If two values have the same hash then they are the same.

I see -- and it can never happen that two values have the same hash but they were created/stored at different point in time, and perhaps we'll need to rollback to a previous version? I suppose not, but it's still good to think about these things when doing these reviews 😀

Reply:
The store is indeed quite dumb now and does deal with rollbacks. Since this is intended to store states I don’t think we even need it at the moment. But it is a good idea to keep an eye on this and possibly extend it.